File: src/main/java/may_2025/ValidBst.java
package may_2025;

public class ValidBst {

    public static void main(String[] args) {
//        TreeNode root = new TreeNode(5);
//        root.left = new TreeNode(1);
//        root.right = new TreeNode(4);
//        root.right.left = new TreeNode(3);
//        root.right.right = new TreeNode(6);

        TreeNode root = new TreeNode(Integer.MAX_VALUE);
//        root.left = new TreeNode(1);
//        root.right = new TreeNode(3);

        ValidBst validBst = new ValidBst();
        boolean result = validBst.isValidBST(root);
        System.out.println(result);
    }

    public boolean isValidBST(TreeNode root) {
        return helper(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }

    private boolean helper(TreeNode node, long min, long max) {
        if (node == null) {
            return true;
        }
        boolean left = helper(node.left, min, node.val);
        boolean right = helper(node.right, node.val, max);

        return left && right && min < node.val && max > node.val;
    }

    static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        public TreeNode(int val) {
            this.val = val;
        }
    }

}

File: python/valid_bst.py
class Solution(object):
    # O(n) time | O(n) space
    def isValidBST(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: bool
        """

        def dfs(node, min, max):
            if not node:
                return True
            
            left = dfs(node.left, min, node.val)
            right = dfs(node.right, node.val, max)

            return left and right and min < node.val and max > node.val
        
        return dfs(root, float("-infinity"), float("infinity"))

        
class TreeNode(object):
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

    def __repr__(self):
        result = str(self.val)
        result += str(self.left) if self.left else ''
        result += str(self.right) if self.right else ''
        return result

root = TreeNode(5)
root.left = TreeNode(1)
root.right = TreeNode(7)
root.right.left = TreeNode(6)
root.right.right = TreeNode(8)

solution = Solution()
print(solution.isValidBST(root))

